html5/js/VideoDecoder.js

```diff
@@ -114,32 +116,46 @@ class XpraVideoDecoder {
   }
 
   queue_frame(packet) {
-    const options = packet[10] || {};
+    return new Promise(async (resolve, reject) => {
+      const options = packet[10] || {};
+      const data = packet[7];
+      const packet_sequence = packet[8];
+
+      if (!this.had_first_key && options["type"] != "IDR") {
+        reject(`first frame must be a key frame but packet ${packet_sequence} is not.`);
+        return;
+      }
 
-    if (!this.had_first_key && options["type"] != "IDR") {
-      this.on_frame_error(packet, "first frame must be a key frame");
-      return;
-    }
+      if (this.videoDecoder.state == "closed") {
+        reject("video decoder is closed");
+        return;
+      }
+      if (this.draining) {
+        reject("video decoder is draining");
+        return;
+      }
 
-    if (this.videoDecoder.state == "closed") {
-      this.on_frame_error(packet, "video decoder is closed");
-      return;
-    }
-    if (this.draining) {
-      this.on_frame_error(packet, "video decoder is draining");
-      return;
-    }
+      this.had_first_key = true;
+      this.decoder_queue.push({ p: packet });
+      const init = {
+        type: options["type"] == "IDR" ? "key" : "delta",
+        data,
+        timestamp: options["frame"],
+      };
+      const chunk = new EncodedVideoChunk(init);
+      this.videoDecoder.decode(chunk);
+
+      let frame_out = this.decoded_frames.filter(p => p[8] == packet_sequence);
+      while (frame_out.length == 0) {
+        // Await our frame
+        await new Promise(r => setTimeout(r, 5));
```

Comment:
Surely there's a way to avoid this "async busy loop"?
(and if not.. there should be an escape hatch? `this.closed` perhaps?)

Reply:
I cant see how to avoid this since the VideoDecoder is using a callback when the frame is decoded.
PR #200 implements a way out (https://github.com/Xpra-org/xpra-html5/pull/200/files#diff-7307ae069666d9ca9815f11190e2e3c02123abdbaaef95a37c5fb7b218ed5029R187)

The frame is either decoded or erroneous.

Reply:
How about `this.videoDecoder.state != "closed"` or `this.initialized`?
(never ending loops make me nervous, in case we somehow find a way to never satisfy the exit condition)

Reply:
I agree and will implement that. We might also need a max-wait variable to break the loop after a while. Will implement this in #200 .